[DEM-NNN]  Load data from subtag

peendebak opened the pull request:
This implements loading all data under a certain subtag. This is common functionality in the QI framework and can reduce the amount of code in methods and scripts.

The implementation is done with a `LazyList`. This is implemented because otherwise all data returned would be loaded into memory which is inefficient or not possible. A use cases for this:
```
data = storage.load_data_from_subtag(tag)
values = [item['field'] for item in data[::10] ]
# do something with values
```

@rhjdjong

rhjdjong commented:
@peendebak Is the usecase you provide a realistic usecase? In other words, do you really need to access every third, or fourth, or tenth item? Or is this just an example of how this _might_ be used?  If you really need to access the items with a stride different from 1, then your implementation is valid. But even in that case, it should be an internal class (so `_LazyList`, with a leading underscore), not something that is offered in the public interface of the module. But if you don't really need something like `[::10]` access, but only want to prevent memory consumption before an item is actually used, then the idiomatic solution is to return a generator expression from `load_data_from_subtag`:
```
subtags = self.list_data_subtags(tag, limit=limit)
return (self.load_data(tag + [subtag]) for subtag in subtags)
```

peendebak commented:
@rhjdjong The use case is a real use case from spin-2. The amount of data stored by now is so much we do not want to process all items.

The generator was my first choice of implementation, but that gives a bad performance since all items are loaded (not all at once, but still they are all loaded). So I changed to the LazyList

I renamed `LazyList` to `_LazySequence` so it is private and better reflects what it actually is.

rhjdjong commented on src/qilib/utils/storage/interface.py:
```diff
@@ -21,11 +21,41 @@
+
+    def __getitem__(self, index: Union[int, slice]) -> Any:
+        if isinstance(index, slice):
+            slice_range = range(index.start or 0, index.stop or len(self), index.step or 1)
```
This can be shortened to `slice_range = range(*index.indices)`

peendebak replied:
Good suggestion, I will update the code. You can use the `Insert a suggestion` from github for such oneliners btw., that makes it easier to update the PR.

rhjdjong commented on src/qilib/utils/storage/interface.py:
```diff
@@ -21,11 +21,41 @@
+    def __getitem__(self, index: Union[int, slice]) -> Any:
+        if isinstance(index, slice):
+            slice_range = range(index.start or 0, index.stop or len(self), index.step or 1)
+            return [self._item_getter(i) for i in slice_range]
```
Here you return an actual list with the components in it. This somewhat defeats the purpose of having lazy evaluation. You could return a generator expression instead:
```
return (self._item_getter(i) for i in slice_range)
```

rhjdjong commented:
@peendebak If the use case you provided is essentially always happening, then you can short-circuit this by providing the indices you want to access as an extra (optional) argument to `load_data_from_subtag`, and return a generator expression that iterates over these indices. This makes the new `_LazySequence` class superfluous. Like:
```
	def load_data_from_subtag(self, tag: TagType, limit: int = 0, indices: Optional[Iterable[int]] = None) -> Iterable[Any]:
		subtags = self.list_data_subtags(tag, limit)
		if indices is None:
			indices = range(len(subtags))
		return (self.load_data(subtags[index]) for index in indices)
```

peendebak commented:
> @peendebak If the use case you provided is essentially always happening, then you can short-circuit this by providing the indices you want to access as an extra (optional) argument to `load_data_from_subtag`, and return a generator expression that iterates over these indices. This makes the new `_LazySequence` class superfluous. Like:
> 
> ```
> 	def load_data_from_subtag(self, tag: TagType, limit: int = 0, indices: Optional[Iterable[int]] = None) -> Iterable[Any]:
> 		subtags = self.list_data_subtags(tag, limit)
> 		if indices is None:
> 			indices = range(len(subtags))
> 		return (self.load_data(subtags[index]) for index in indices)
> ```

This would require the user to know the number of elements in the query beforehand (to construct the `indices`). There are ways to get that working, but it would require changes in `qubits-control`. Since we have not decided on a design for LLS2 I don't want to add changes that are not backward compatible at this moment.

rhjdjong commented:
> > @peendebak If the use case you provided is essentially always happening, then you can short-circuit this by providing the indices you want to access as an extra (optional) argument to `load_data_from_subtag`, and return a generator expression that iterates over these indices. This makes the new `_LazySequence` class superfluous. Like:
> > ```
> > 	def load_data_from_subtag(self, tag: TagType, limit: int = 0, indices: Optional[Iterable[int]] = None) -> Iterable[Any]:
> > 		subtags = self.list_data_subtags(tag, limit)
> > 		if indices is None:
> > 			indices = range(len(subtags))
> > 		return (self.load_data(subtags[index]) for index in indices)
> > ```
> 
> This would require the user to know the number of elements in the query beforehand (to construct the `indices`). There are ways to get that working, but it would require changes in `qubits-control`. Since we have not decided on a design for LLS2 I don't want to add changes that are not backward compatible at this moment.

Good argument for keeping the _LazySequence class.
